### DISCUSSION TITLE
[SOLVED] How likely is a switch to TypeScript?

### DISCUSSION THREAD

**Discussion by trusktr** [pulsar-edit/.github#139] [Core Application & Packages]
Just wondering.

I've been working on compiling TypeScript to WebAssembly. I've also written the first Custom Elements (Web Components) ever compiled to WebAssembly. I think it could be interesting to compile the editor to WebAssembly too!

**Comment by Daeraxa** [comment]
Unlikely. Aside from some members of the team flat out not liking TS, we are currently trying to get back to pure JS as much as possible by removing all the existing CoffeeScript (decaffeinating). To me it makes sense, our goal is to be a hackable text editor and TS just raises that barrier of entry a little bit higher.
Currently no real compilation or transpilation steps are needed to make small changes to the underlying code and many have expressed a desire to get to a point where not even an editor reload is needed and changes can be reflected in real time.

**Reply by trusktr** [reply]
Yeah, even if the editor is TypeScript, people can still hack it in plain JS (and get nice intellisense while at it).

> expressed a desire to get to a point where not even an editor reload is needed and changes can be reflected in real time.

Type defs can be done entirely through JSDoc comments. Type checking can even still be strict while writing JS with comments. It's a little more verbose, but does satisfy the need to run plain JS without a build step. If there's a build though, then its moot, TS gives terser type syntax. (looking to support JS+JSDoc-to-Wasm compile in my project too)

---

If the team is averse to TypeScript, then yeah this idea basically ends here.

**Reply by meadowsys** [reply]
> expressed a desire to get to a point where not even an editor reload is needed and changes can be reflected in real time

Typescript / other types of transpilation aren't really a barrier to this happening. Its just that you would have tsc in watch mode on the side, or whatever other transpilation tool, and the js files are still changing on the filesystem, just its a transpiler changing it instead of hitting save in a text editor.  To the editor / whatever is doing the updating, it looks the same.

Personally, I am okay with typescript, but agree that the editor should be kept in in javascript. Not sure if the editor has typings in the jsdoc comments or not, and not sure where the discussion is at for that. I don't speak for the team, but I think it would be a nice to have, for the better autocomplete and extra documentation (don't turn on type checking). JSDoc is a little clunky on the types side, and there are a few typescript typings that aren't supported, but they are relatively minor edge cases.
